Add table enhancements and owner validation

radchukd opened the pull request:
Closes #43, #44, #45

radchukd committed:
make oldOwnerAddress not changeable

radchukd committed:
Add Transactions context, add table sorting and pagination via react-table

radchukd committed:
Add Jazzicon component
Add OwnersContext
Add owner validation
Add better wallet loading

radchukd committed:
Remove toast and modal handling from owners/transactions contexts

radchukd committed:
Add changeThreshold to OwnersContext

radchukd commented:
PR is a bit big, here's a summary:

**1. Contexts**
I added 2 contexts - `OwnersContext` and `TransactionsContext`, which provide data respectively and handle all the smart contract interactions, in result:
- removed code repetition;
- added single source of truth with easier access to it;
- added separation between data and UI;
- made loading more predictable;
- made table manipulation possible.

**2. Table**
I added _react-table_ library (MIT) to handle sorting and pagination in the transactions table.

**3. Other**
I changed the `ReplaceOwnerModal` according to #43.
I did everything in #44 (added jazzicon component, owner validation, renaming).

radchukd committed:
Make loading more reactive

radchukd committed:
Remove unnecessary fetches

apbendi reviewed (changes requested):
Hey @radchukd, looking great overall! A couple things I've noticed:

* Sorting seems to be broken. Clicking the headers makes the table flicker, but the sorting order/column does not change
* The "Show N" dropdown works well, but it should *only show up if the wallet has more than 5 txs to display*. Before that point, it should be hidden.

apbendi commented:
There also seems to be some kind of state bug occurring when navigating between wallets. Here's a brief example:

![ogg-state-bug](https://user-images.githubusercontent.com/1479802/136311691-83d4068f-a8a2-478a-bf89-d1aee9dfede9.gif)

radchukd committed:
Add owner count condition for threshold changing
Add txs count condition for txs count changing
Fix sorting table updates
Fix txs update from transtaction row

radchukd commented:
Now also closes #47

apbendi commented:
Hey @radchukd, sorting and threshold button changes looking good! However, I am still seeing the state issues in the UI when switching between wallets. Note how when I start with an empty wallet, switch to one with lots of txs, then navigate back to the empty wallet, I see empty tables displayed. When I refresh, it looks the way it should:

![ogg-state-bug-more](https://user-images.githubusercontent.com/1479802/136551459-871a51a7-0d0a-4351-a3dc-4f34757c700b.gif)

radchukd committed:
Move Wallet providers to wallet page

apbendi commented:
@radchukd looks like there's a regression on the "Change Threshold" form validation. Not accepting anything other than 1, even when there are multiple owners:

![change-threshold-bug](https://user-images.githubusercontent.com/1479802/136612460-7fb4f16e-1b59-448b-9768-10970dd5c7ba.gif)

radchukd committed:
Revert "Move Wallet providers to wallet page"

This reverts commit 89aebf40ca4cffddbbd71f1ab2c402b8fdfbb3ce.

radchukd committed:
Fix switching to a wallet without transactions

radchukd commented:
Oops, the problem was in a totally different place, sorry. I reverted the previous changes and pushed a new commit, should be good now.

apbendi reviewed (commented):
Hey @radchukd, the state issues I was seeing when navigating between wallets seems to be resolved, though I'm now seeing a similar issue when creating a new tx. Check it out:

![new-tx-bug](https://user-images.githubusercontent.com/1479802/138317212-a16c9a06-6b23-43aa-a325-b446967bdf56.gif)

radchukd committed:
Fix new tx submission

radchukd commented:
@apbendi should be good now.
